### ISSUE TITLE
Istio istio_request_total counter goes to 0 after istio-proxy restart

### ISSUE THREAD

**Issue by mcervar** [istio/istio#44193]
Hi

After days and days of investigation, I simply didn't find any clue how to make istio_request_total counter persistent across istio-proxy container restarts.
We are using it to track how many times client opened specific url_path on specific service.

Tried mount persistent volume to /var/lib/istio/data but no effect.
Is there a way to make that istio_request_total counter preserve state across container restart?

Thx in advance

**Comment by howardjohn** [comment]
The state is stored in Prometheus

**Comment by mcervar** [comment]
Yes, history data is stored, but to be more precise.

I am following number of request towards specific app API calls. 
Those metrics are generated on istio-proxy aka envoy and collected by Prometheus. 

For test, restarted istio-proxy, state of the metric was 0. Generated 20 API request through istio, checked status on Prometheus, it says 20 request were made. So the number of request made was correctly recorded.

After that I restarted istio-proxy container. Metric is reseted to 0. So history data shows that metric dropped from 20 to 0.
Restart of Prometheus doesnt have effect on data since it has persistent storage.

So question is how to properly track numbers of request without metric being reset because of restart? Or have to perserve state of metric on istio container across restarts?

**Comment by zirain** [comment]
are you running istio-proxy on VM?

**Comment by mcervar** [comment]
In this case istio-proxy represents container inside istio-gateway pods in K8S cluster. Prometheus collects metrics from those pods.

**Comment by zirain** [comment]
I think you should figure out how Prometheus work first.
Istio does nothing special.

**Comment by mcervar** [comment]
I am working daily with Prometheus for past 3 years XD

Once again, prometheus works good and properly collect metrics and persist them on disk, this is not the issue.
Issue is that istio-proxy container resets the value of metric to 0, then prometheus takes it.

So for example if total number of istio_request_total metrics is 300, after restart of istio it will be 0, because metric will be reseted to 0 on istio-proxy side, and prometheus will just collect it

**Comment by zirain** [comment]
@mcervar do you means suitation like following:
```
NAMESPACE            NAME                                          READY   STATUS    RESTARTS       AGE
default              httpbin-797587ddc5-4kj2v                      2/2     Running   1 (9m5s ago)   16m
```

**Comment by mcervar** [comment]
Yes, for example, you can collect istio_request_total metrics on your httpbin pod, this metrics is type counter and it will increase based on the number of request made towards httpbin pod, but if you restart it, istio_request_total metrics will be reseted to 0

**Comment by zirain** [comment]
maybe `__meta_kubernetes_pod_container_id` can help you, https://github.com/prometheus/prometheus/pull/11844

**Comment by mcervar** [comment]
I already tried creating statefulset pods (and even mounting persistent storage instead of empty folder volume for istio/envoy data directory), which will not change name, but metrics are still reseted on restart.

Suggested metric helps to differentiate metrics between containers, which is not the issue in my case :/

**Comment by zirain** [comment]
can you try a prometheus demo like this: https://github.com/zirain/playground/blob/master/istio/prometheus.yaml

**Comment by mcervar** [comment]
I am already using prometheus, here is  example how it acts, this is grafana panel using prometheus datasource

![image](https://user-images.githubusercontent.com/47822784/232518854-af8ccf3b-34e6-481c-ae9e-56125a7bebe3.png)

**Comment by zirain** [comment]
is that dashboard using `increase` or `rate`?

**Comment by mcervar** [comment]
@zirain thx a lot for suggestion! Increase solved the issue with counter reset on my grafana :D
